processing/src/main/java/org/apache/druid/segment/data/CompressionStrategy.java

```diff
@@ -344,6 +360,81 @@ public ByteBuffer compress(ByteBuffer in, ByteBuffer out)
     }
   }
 
+  public static class ZstdCompressor extends Compressor
+  {
+    private static final ZstdCompressor DEFAULT_COMPRESSOR = new ZstdCompressor();
+
+    @Override
+    ByteBuffer allocateInBuffer(int inputSize, Closer closer)
+    {
+      ByteBuffer inBuffer = ByteBuffer.allocateDirect(inputSize);
+      closer.register(() -> ByteBufferUtils.free(inBuffer));
+      return inBuffer;
+    }
+
+    @Override
+    ByteBuffer allocateOutBuffer(int inputSize, Closer closer)
+    {
+      ByteBuffer outBuffer = ByteBuffer.allocateDirect((int) Zstd.compressBound(inputSize));
+      closer.register(() -> ByteBufferUtils.free(outBuffer));
+      return outBuffer;
+    }
+
+    @Override
+    public ByteBuffer compress(ByteBuffer in, ByteBuffer out)
+    {
+      int position = in.position();
+      out.clear();
+      long sizeNeeded = Zstd.compressBound(in.remaining());
+      if (out.remaining() < sizeNeeded) {
+        throw new RuntimeException("Output buffer too small, please allocate more space. " + sizeNeeded + " required.");
+      }
+      Zstd.compress(out, in, Zstd.maxCompressionLevel());
+      in.position(position);
+      out.flip();
+      return out;
+    }
+  }
+
+  public static class ZstdDecompressor implements Decompressor
+  {
+    private static final ZstdDecompressor DEFAULT_COMPRESSOR = new ZstdDecompressor();
+
+    @Override
+    public void decompress(ByteBuffer in, int numBytes, ByteBuffer out)
+    {
+      out.clear();
+      // some tests don't use dbb's and zstd jni doesn't allow for non-dbb byte buffers.
```

Comment:
hmm, looking around it seems like it should the places we use compression should only be using direct buffers, what do you think about making this an error condition instead and just skipping the tests that use heap buffers?

If we find out that it is in fact needed, then the `cloneBuffer` code needs to be modified to call `ByteBufferUtils.free` after it is done with the direct buffer so that we do not leak memory.

Reply:
yeah, everyone should use DBB, but I believe when running the following test: It failed with: 

```[ERROR] testSanityWithSerde[2](org.apache.druid.segment.data.CompressedColumnarIntsSupplierTest)  Time elapsed: 0 s  <<< ERROR!
     java.lang.IllegalArgumentException: srcBuff must be a direct buffer
     	at com.github.luben.zstd.ZstdDecompressCtx.decompressDirectByteBuffer(ZstdDecompressCtx.java:110)
     	at com.github.luben.zstd.Zstd.decompressDirectByteBuffer(Zstd.java:434)
     	at org.apache.druid.segment.data.CompressionStrategy$ZstdDecompressor.decompress(CompressionStrategy.java:407)
     	at org.apache.druid.segment.data.DecompressingByteBufferObjectStrategy.fromByteBuffer(DecompressingByteBufferObjectStrategy.java:53)
     	at org.apache.druid.segment.data.DecompressingByteBufferObjectStrategy.fromByteBuffer(DecompressingByteBufferObjectStrategy.java:28)
     	at org.apache.druid.segment.data.GenericIndexed$BufferIndexed.bufferedIndexedGet(GenericIndexed.java:490)
     	at org.apache.druid.segment.data.GenericIndexed$3.get(GenericIndexed.java:645)
     	at org.apache.druid.segment.data.CompressedColumnarIntsSupplier$CompressedColumnarInts.loadBuffer(CompressedColumnarIntsSupplier.java:321)
     	at org.apache.druid.segment.data.CompressedColumnarIntsSupplier$CompressedColumnarInts.get(CompressedColumnarIntsSupplier.java:312)
     	at org.apache.druid.segment.data.CompressedColumnarIntsSupplierTest.assertIndexMatchesVals(CompressedColumnarIntsSupplierTest.java:289)
     	at org.apache.druid.segment.data.CompressedColumnarIntsSupplierTest.testSanityWithSerde(CompressedColumnarIntsSupplierTest.java:175)
```

I don't know what is the right solution here, I agree what I wrote feels a bit dirty.  I'll defer to you here on how you think we should sort this out..

Reply:
I think the right way is to probably make it be an exception to use this compression strategy without direct buffers, and modify the test to use direct buffers (or to skip this compression strategy).

Reply:
It seems like it mostly would be replacing `ByteBuffer.wrap` with a direct buffer, and adding an `@After` to release them, unless I'm missing something?

If you'd rather not modify the tests (though it seems like they would probably be better tests using direct buffers since it more closely reflects what would be happening in practice), I would be ok with instead modifying the code to call `ByteBufferUtils.free` on the cloned direct buffers, so that in case I missed something and we ever do use heap buffers in production code they are at least not leaked.

Reply:
I will send you the test where I did the `ByteBuffer.wrap`, I did that and there are still internal heap byte buffers created somewhere.  I can look into it further in another PR but I agree lets get that resolved as well.  Here is what I did https://gist.github.com/churromorales/9b41f7b62e4debb21a94a2c357a811a5 (also removing the conversion in CompressionStrategy).  Perhaps I missed something?  

As far as freeing the byte buffers, if we clone the buffer, we should only free the input clone buffer right?  I would assume the it would be the caller's responsibility to free the cloned output buffer as they are reading that.  I just want to make clear that is what you want and that I am not missing anything there. 

Thank you again for the review.

Reply:
Sorry for the delay, but I have a better idea I think on what to do... how about instead of cloning into a direct buffer, we just use byte arrays and `Zstd.decompressByteArray` and copy the output byte array to the output buffer.

```java
    @Override
    public void decompress(ByteBuffer in, int numBytes, ByteBuffer out)
    {
      out.clear();
      // fall back to heap byte arrays if both buffers are not direct
      if (!in.isDirect() || !out.isDirect()) {
        byte[] inputBytes = new byte[numBytes];
        in.get(inputBytes);
        try (final ResourceHolder<byte[]> outputBytesHolder = CompressedPools.getOutputBytes()) {
          final byte[] outputBytes = outputBytesHolder.get();
          int decompressedBytes = (int) Zstd.decompressByteArray(outputBytes, 0, outputBytes.length, inputBytes, 0, numBytes);
          out.put(outputBytes, 0, decompressedBytes);
          out.flip();
        }
      } else {
        int decompressedBytes = (int) Zstd.decompressDirectByteBuffer(
            out,
            out.position(),
            out.remaining(),
            in,
            in.position(),
            numBytes
        );
        out.limit(out.position() + decompressedBytes);
      }
    }
```
I tested it locally and it seems to work.


Also make sure any modifications done to tests in this PR that might allocate direct buffers are releasing them, see #12521, which also looks like has caused a conflict with this PR. I'll have a scan and see if I can spot anything too.

Reply:
@clintropolis Thanks for the review.  I agree, just having a byte array is the easiest thing.  I fixed the checkstyle issues and as for the PR modifications from #12521 , you want me to rebase and force push, or merge main into my branch?  How do you like to do this such that I can fix the `CompressionStrategyTest`?

Reply:
There was a reason for modifying the tests.  The only thing that was calling compress with a non-dbb was that `CompressionStrategyTest`.  We don't support a non-dbb `compression` in zstandard right now.  The only thing that changed about the tests was not to test the non-dbb path.  Which I thought was acceptable, because in the end we should never be calling this code with any other buffer than a direct.  

I can fix the rest of the open items.  I will wait to hear from you about what you want to do with the CompressionStrategyTest.  If it is worthwhile, I'll have to do buffer copies for the compress like what was done in the Decompressor. 

Again thank you for the review and making my life easier debugging the CICD issues.

Reply:
Ah I see now, yeah that makes sense :+1: I think I'm ok with leaving the test like this. 

I guess we could add the `byte[]` version fallback of `compress` just in case anything does try to call it that we didn't find, though I _think_ all compression uses the 'allocate' methods on `CompressionStrategy` so its probably not necessary.